### PR TITLE
Convert BIDS dataset to the MSD format

### PR DESCRIPTION
**Pull request by naga-karthik**
This PR adds a template for converting any BIDS dataset to a Medical Segmentation Decathlon (MSD) format to be used with MONAI's data loaders. Description and usage examples are provided within the script.

### CODE REVIEW

File: bids_to_msd_dataset_conversion.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+import json
+import pandas as pd
+from tqdm import tqdm
+import numpy as np
+import argparse
+import nibabel as nib
+from loguru import logger
+from sklearn.model_selection import train_test_split
+import utils_dataset_conversion as utils
+import glob
```
Comment (valosekj) [inline bids_to_msd_dataset_conversion.py:54]: Maybe you could add `requirements.txt` file?

File: bids_to_msd_dataset_conversion.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+                        if contrast_suffixID == args.common_label_contrast:
+                            label_file = os.path.join(root, 'derivatives', 'labels', subjectID, datatype, utils.add_suffix(filename, args.label_suffix))
+
+                        # # TODO: check if there are missing contrasts
+
+                    elif args.include_contrasts is None:
+                        # assuming all contrasts have to be included
+                        image_file = os.path.join(root, subjectID, datatype, filename)
+                        if contrast_suffixID == args.common_label_contrast:
+                            label_file = os.path.join(root, 'derivatives', 'labels', subjectID, datatype, utils.add_suffix(filename, args.label_suffix))
```
Comment (valosekj) [inline bids_to_msd_dataset_conversion.py:175]: `'derivatives', 'labels'` is hardcoded several times across the script. It would be better to define it as a global variable at the beginning of the script, e.g.:

```
PATH_DERIVATIVES = os.path.join('derivatives', 'labels')
```

This will allow easy modification in case the user uses, for example, `derivatives/manual_labels` instead of `derivatives/labels`

File: bids_to_msd_dataset_conversion.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+    params[name] = temp_list
+
+
+# get the number of training, validation and test samples depend on the depending on the grouping/sessions/contrasts chosen
+params["numTraining"] = len(params["training"])
+params["numValidation"] = len(params["validation"])
+params["numTest"] = len(params["test"])
+
+final_json = json.dumps(params, indent=4, sort_keys=True)
+jsonFile = open(os.path.join(args.path_out, f"dataset.json"), "w")
```
Comment (valosekj) [inline bids_to_msd_dataset_conversion.py:288]: If `args.path_out` does not exist, the script fails. There are two possible workarounds:
1. exit the script if the `args.path_out` does not exist
2. check if `args.path_out` exists, if not, create it

File: bids_to_msd_dataset_conversion.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+params["description"] = "My awesome task"       # TODO: Add the name of the data
+params["labels"] = {
+    "0": "background",
+    "1": "sc-lesion"        # TODO: Define the classes to be be segmented
+    }
+params["license"] = "xxx"
+params["modality"] = {
+    "0": "MRI"
+    }
+params["name"] = "sci-zurich"       # TODO: Add the name of the dataset
```
Comment (valosekj) [inline bids_to_msd_dataset_conversion.py:125]:
```suggestion
params["name"] = "dataset-name"       # TODO: Add the name of the dataset
```

File: bids_to_msd_dataset_conversion.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+import nibabel as nib
+from loguru import logger
+from sklearn.model_selection import train_test_split
+import utils_dataset_conversion as utils
+import glob
+
+parser = argparse.ArgumentParser(description='Code for creating datalists according to the MSD format.')
+
+parser.add_argument('--seed', default=42, type=int, help="Seed for reproducibility")
+parser.add_argument('--path_data', default='./', required=True, type=str,
```
Comment (valosekj) [inline bids_to_msd_dataset_conversion.py:59]: I think that usually `-` is commonly used over `_` in input args names, e.g., `--path-data`

File: bids_to_msd_dataset_conversion.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+    "validation": val_subjects,
+    "test": test_subjects
+}
+
+# loop through the training, validation and test splits and create dictionaries that contain the 
+# (image, label) pairs for each subject
+for name, subs_list in subjects_dict.items():
+
+    temp_list = []
+    for subject_no, subject in enumerate(tqdm(subs_list, desc='Loading Volumes')):
```
Comment (valosekj) [inline bids_to_msd_dataset_conversion.py:140]: Now, `'Loading Volumes'` is printed 3x times to CLI (for train/valid/test). At first look, it is not clear that the first progress bar is related to the training fraction etc. Maybe, `desc` could be modified:
```
desc=f'Loading {name} Volumes'
```

File: bids_to_msd_dataset_conversion.py (hunk 5 above)
Comment (naga-karthik) [reply bids_to_msd_dataset_conversion.py:59]: If we use `--path-data` then we can't use the same variable in the script elsewhere right? Like, with `--path_data`, I can still use `args.path_data` but if I change it to `--path-data` then does `args.path_data` still work?

Comment (naga-karthik) [reply bids_to_msd_dataset_conversion.py:59]: Just checked that it works! 🤯 I really didn't know this!

File: bids_to_msd_dataset_conversion.py (hunk 6 above)
Comment (naga-karthik) [reply bids_to_msd_dataset_conversion.py]: Added in https://github.com/ivadomed/templates/pull/1/commits/bf260b414fbd197f56b793c394cf5b5b7108a65a!

File: bids_to_msd_dataset_conversion.py (hunk 5 above)
Comment (naga-karthik) [reply bids_to_msd_dataset_conversion.py]: Thanks for this suggestion! Changed in https://github.com/ivadomed/templates/pull/1/commits/bf260b414fbd197f56b793c394cf5b5b7108a65a!

**Comment by valosekj** [comment]
Tested [Single Session Single Contrast](https://github.com/ivadomed/templates/blob/b1129df5b83a5edd1ce58d71509a1c641e791e0f/bids_to_msd_dataset_conversion.py#L25) case on `canproco` dataset.

Since one site has STIR contrast and the remaining sites have PSIR contrast, I had to run the command twice:

```
python bids_to_msd_dataset_conversion.py --path-data canproco --path-out ./ --split 0.6 0.2 0.2 --label-suffix _lesion-manual --include-contrasts STIR
python bids_to_msd_dataset_conversion.py --path-data canproco --path-out ./ --split 0.6 0.2 0.2 --label-suffix _lesion-manual --include-contrasts PSIR
```

Is this the correct usage in this scenario when the contrast differs among sites?

Also, for PSIR contrast which is available for four sites, does the script split the subjects equally among sites for test/valid/train fractions?

**Comment by naga-karthik** [comment]
> Is this the correct usage in this scenario when the contrast differs among sites?

Yes, in general, if certain sites have only certain contrasts (assuming mutually exclusive) then we would have to: (i) run the commands separately, (ii) get the corresponding json datalists, and (iii) merge the jsons to get a single datalist for the entire dataset (if required)

> Also, for PSIR contrast which is available for four sites, does the script split the subjects equally among sites for test/valid/train fractions?

No, I just checked by running one of your commands for PSIR earlier. It seems that the splitting between sites is unequal for now.

**Comment by jcohenadad** [comment]
> This PR adds a template for converting any BIDS dataset to a Medical Segmentation Decathlon (MSD) format to be used with MONAI's data loaders.

I am skeptical about the relevance/strategical choice of wanting to convert our BIDS dataset for training, instead of making the MONAI data loader compatible. Two main arguments:
- Adding a dataset conversion means more overhead when the source dataset is updated (ie: needs to re-convert). Moreover, the conversion is prone to errors because the converted data is not version tracked anymore (ie: our BIDS datasets are assigned a git hashtag, which is lost upon conversion).
- I do not think that BIDS loader is intrinsically incompatible with BIDS format. Eg: In [this project](https://github.com/jcohenadad/model-seg-ms-mp2rage-monai/blob/58e8f096343bca618b6bc137e183eb9cb186f1ab/train_wandb.py#L52-L101) BIDS dataset is loaded, into a MONAI tutorial. The rest of the tutorial code is unchanged.

**Comment by naga-karthik** [comment]
Hey @jcohenadad, note that these template scripts **do not** convert the existing the BIDS datasets into the MSD format. We are indeed making it MONAI data loader compatible. This is explicitly mentioned in the docstring of the script [here](https://github.com/ivadomed/templates/blob/nk/bids-to-msd/bids_to_msd_dataset_conversion.py#L4-L6). 

How we're making it MONAI compatible is by pointing to the corresponding image and label files in the existing BIDS dataset (which is what your [function](https://github.com/jcohenadad/model-seg-ms-mp2rage-monai/blob/58e8f096343bca618b6bc137e183eb9cb186f1ab/train_wandb.py#L52-L71) also does). Here's an [example](https://github.com/ivadomed/templates/blob/nk/bids-to-msd/bids_to_msd_dataset_conversion.py#L37-L53) of the dataset.json looks like after running this script. 

> Moreover, the conversion is prone to errors because the converted data is not version tracked anymore

As I clarified before, since we're not re-structuring the bids dataset, we don't worry about the version tracking. The script using whichever bids dataset the points it to (whether it be an earlier version or the latest version)

> Eg: In [this project](https://github.com/jcohenadad/model-seg-ms-mp2rage-monai/blob/58e8f096343bca618b6bc137e183eb9cb186f1ab/train_wandb.py#L52-L101) BIDS dataset is loaded, into a MONAI tutorial.

An issue I see with this function is that it only pairs one image and its corresponding label. However, we have more cases where we want to combine multiple sessions/contrasts etc. For example, [this](https://github.com/ivadomed/templates/blob/nk/bids-to-msd/bids_to_msd_dataset_conversion.py#L66-L84) is how we can combine multiple contrasts in a bids dataset using this script. 

Bottomline is that the template script is essentially doing what your function is doing as well, except that it gives more options to combine the (image, label) pairs in different ways depending what the user wants.

**Comment by jcohenadad** [comment]
Excellent! Thank you for the clarification @naga-karthik and I'm glad to see our visions are aligned. 

The fact that I did not understand the purpose of the script probably suggests that its docstring needs to be updated. Notably, starting the description with "Template for converting a BIDS dataset to the Medical Segmentation Decathlon format." is misleading. I suggest changing it for: "Convert the JSON dataset description file from BIDS to Medical Segmentation Decathlon (MSD)."

Also, I would stay away from the words "Template script". this is neither a script (but a function), and also it is not a template in a sense that it should be edited and customized by other users. On the contrary, it is a function, that will be called by a custom training script such as [the one I linked](https://github.com/jcohenadad/model-seg-ms-mp2rage-monai/blob/58e8f096343bca618b6bc137e183eb9cb186f1ab/train_wandb.py#L52-L101).

**Comment by naga-karthik** [comment]
I definitely had some trouble explaining the exact purpose of this file, hence, thank you for the suggestions! This has been updated in commit https://github.com/ivadomed/templates/pull/1/commits/e66ff50ae1c4b2cceb2dcc797c469f6cee6f6289.
